binary_search_tree/binary_search_tree.py

```diff
@@ -12,21 +12,59 @@ def __init__(self, value):
 
     # Insert the given value into the tree
     def insert(self, value):
-        pass
+        if value < self.value:
+            # check if self.left is a valid node
+            if self.left:  # is Not None / is there a node here
+                self.left.insert(value)
+            # base case - the left side is empty
+            else:
+                # we've found our valid parking spot
+                self.left = BinarySearchTree(value)
+        # otherwise, value >= self.value
+        else:
+            if self.right:
+                self.right.insert(value)
+            else:
+                self.right = BinarySearchTree(value)
 
     # Return True if the tree contains the value
     # False if it does not
     def contains(self, target):
-        pass
+        # if the value is the target, we're done!
+        if self.value == target:
+            return True
+        if target >= self.value and self.right:  # if the target is greater than or equal to AND self.right is valid
+            # call the function again recursively on the right (greater than) side
+            return self.right.contains(target)
+        elif target < self.value and self.left:
+            # call the function again recursively on the left (less than) side
+            return self.left.contains(target)
+        else:
+            return False  # not here
 
     # Return the maximum value found in the tree
     def get_max(self):
-        pass
+        # we will only work with the right side, as the left side can never be the max num
+        if self.right:
+            return self.right.get_max()
+        else:
+            return self.value
 
     # Call the function `cb` on the value of each node
     # You may use a recursive or iterative approach
     def for_each(self, cb):
-        pass
+        if self.right and self.left:
+            cb(self.value)
+            self.right.for_each(cb)  # call it again on the items to the right
+            self.left.for_each(cb)  # call it again on the items to the left
+        elif self.left:  # no more to the right, but still more to left
+            cb(self.value)
+            self.left.for_each(cb)
+        elif self.right:  # no more to the left, but still more to the right
+            cb(self.value)
+            self.right.for_each(cb)
+        else:
+            cb(self.value)
```

Comment:
There's a lot of redundancy in this implementation. 

How could you trim off two of these conditionals? Is it necessary to check for `self.left and self.right` if you're already checking for them each individually? Is it necessary for these checks to be exclusive? (i.e. `if/elif`)

`cb()` is called in every conditional, how can that be made more dry?